### PR TITLE
#21 Cache

### PR DESCRIPTION
**Pull request by ammaratef45**
a PR for #21 
Added flask cache with Redis.
Cashing templates.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#164](https://github.com/timelesslounge/timelessis/issues/164) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#164](https://github.com/timelesslounge/timelessis/issues/164) is assigned to @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/e00a8113-edcd-42ed-a094-9847d2a00ab3); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
Job `gh:timelesslounge/timelessis#164` already assigned to @vladarefiev, can't assign to @amihaiemil

**Comment by 0crat** [comment]
Job `gh:timelesslounge/timelessis#164` already assigned to @vladarefiev, can't assign to @vladarefiev

File: timeless/templates/auth/activate.html [hunk 1]
```diff
@@ -1,5 +1,7 @@
 {% extends 'base.html' %}
 
+{% cache 600 %}
```
Comment (vladarefiev) [inline timeless/templates/auth/activate.html]: @ammaratef45 I don't know what exactly `activate` template here. but I guess we should not cache it

File: timeless/__init__.py [hunk 2]
```diff
@@ -19,11 +19,14 @@
...
 from timeless.restaurants.table_shapes import views as table_shapes_views
 
+from flask_cache import Cache
+
 
 def create_app(config):
     """Creates a new Timeless webapp given a config class"""
     app = Flask(__name__, instance_relative_config=True)
     app.config.from_object(config)
+    cache = Cache(app, config={'CACHE_TYPE': 'redis'})
```
Comment (vladarefiev) [inline timeless/__init__.py]: @ammaratef45 At least pls update README how to set up redis.
But actually we have a Dockerfile and I prefer to leave here puzzle to set up redis to `docker-compose.yaml`.

**Comment by vladarefiev** [comment]
@ammaratef45 just a few comments pls

**Comment by ammaratef45** [comment]
@vladarefiev Thanks, I implemented them please take a look.

**Comment by vladarefiev** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@ammaratef45 I guess we need to do something on rultor

**Comment by ammaratef45** [comment]
@emilianodellacasa I read the log, nothing related to the changes
I wonder 😕

**Comment by ammaratef45** [comment]
@emilianodellacasa Hmm, Rultor probably can't find `flask_cache` 
I don't know how Rultor merges but, I added the library to the requirements file so is that enough?

**Comment by emilianodellacasa** [comment]
@ammaratef45 take a look at this https://github.com/thadeusb/flask-cache/issues/188

**Comment by ammaratef45** [comment]
@emilianodellacasa in the commit d7cfdaaadae7c3e41cfa92b1e94730e27e51bb2d the fix in the comments were implemented, and it worked locally

**Comment by emilianodellacasa** [comment]
@ammaratef45 so why rultor cannot?

**Comment by ammaratef45** [comment]
@emilianodellacasa I cleared my environment then followed the rultor script and reproduced,
Working on it

**Comment by ammaratef45** [comment]
@emilianodellacasa I solved the problem.
However, Rultor will keep failing because it should install and run Redis first.
We need to change the script first.

**Comment by vladarefiev** [comment]
@ammaratef45 add redis installation to `.rultor.yml` `install` section

**Comment by ammaratef45** [comment]
@vladarefiev should do this in a separate PR?

**Comment by vladarefiev** [comment]
@ammaratef45 I don't know, let's ask @emilianodellacasa 
BTW we can't merge with PR without installed redis.

**Comment by ammaratef45** [comment]
I will prepare it anyway

**Comment by ammaratef45** [comment]
Should be fine now

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@ammaratef45 look at rultor

**Comment by ammaratef45** [comment]
@emilianodellacasa 
I noticed that rultor changed `sudo make install` to `sudo -i make install` and that what causes the problem.
Do you know how I can trace this?

**Comment by emilianodellacasa** [comment]
@ammaratef45 try to put the same code in an external script

**Comment by ammaratef45** [comment]
I used [This](https://gist.github.com/ammaratef45/eaef2984596bccc80b1800f92efa1e4c) script and run it locally, Works fine.
It contains the same commands I gave to rultor.

**Comment by emilianodellacasa** [comment]
@ammaratef45 good, then add the script to the repo and call it from rultor as we do for other scripts

**Comment by ammaratef45** [comment]
@emilianodellacasa try now

**Comment by emilianodellacasa** [comment]
@ammaratef45 while you are at it, could you please move your script and install_db.sh under scripts/install dir and change all references in rultor? It would be cleaner this way, thanks!

**Comment by ammaratef45** [comment]
> @ammaratef45 while you are at it, could you please move your script and install_db.sh under scripts/install dir and change all references in rultor? It would be cleaner this way, thanks!

Done, let's try

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by ammaratef45** [comment]
@emilianodellacasa Rultor is stuck at the `redis-server` command 
I edited the command  ffefc44

**Comment by emilianodellacasa** [comment]
@rultor stop

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@ammaratef45 you have single quotes on your code, please fix

**Comment by ammaratef45** [comment]
@emilianodellacasa `checkstyle.py` script is new 😅

**Comment by emilianodellacasa** [comment]
@ammaratef45 there are conflicts now, please fix

**Comment by ammaratef45** [comment]
@emilianodellacasa Done
Merge and I will pray 😆

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by ammaratef45** [comment]
> > @rultor merge
> 
> @ammaratef45 @emilianodellacasa Oops, I failed. You can see the full log [here](http://www.rultor.com/t/17774-462167943) (spent 6min)
> 
> ```
> timeless/roles/views.py:63:0: C0111: Missing function docstring (missing-docstring)
> ************* Module timeless.access_control.authorization
> timeless/access_control/authorization.py:14:0: C0301: Line too long (105/100) (line-too-long)
> timeless/access_control/authorization.py:19:0: C0305: Trailing newlines (trailing-newlines)
> timeless/access_control/authorization.py:1:0: C0111: Missing module docstring (missing-docstring)
> timeless/access_control/authorization.py:4:0: W1113: Keyword argument before variable positional arguments list in the definition of is_allowed function (keyword-arg-before-vararg)
> ************* Module timeless.access_control.owner_privileges
> timeless/access_control/owner_privileges.py:1:0: C0111: Missing module docstring (missing-docstring)
> timeless/access_control/owner_privileges.py:4:0: W1113: Keyword argument before variable positional arguments list in the definition of has_privilege function (keyword-arg-before-vararg)
> timeless/access_control/owner_privileges.py:4:0: W0613: Unused argument 'kwargs' (unused-argument)
> timeless/access_control/owner_privileges.py:15:0: W1113: Keyword argument before variable positional arguments list in the definition of __location_access function (keyword-arg-before-vararg)
> timeless/access_control/owner_privileges.py:15:0: W0613: Unused argument 'args' (unused-argument)
> ************* Module timeless.access_control.views
> timeless/access_control/views.py:1:0: C0111: Missing module docstring (missing-docstring)
> ************* Module timeless.access_control.methods
> timeless/access_control/methods.py:10:0: C0305: Trailing newlines (trailing-newlines)
> timeless/access_control/methods.py:1:0: C0111: Missing module docstring (missing-docstring)
> ************* Module timeless.access_control.__init__
> timeless/access_control/__init__.py:26:0: C0305: Trailing newlines (trailing-newlines)
> ************* Module timeless.restaurants.models
> timeless/restaurants/models.py:146:0: C0305: Trailing newlines (trailing-newlines)
> timeless/restaurants/models.py:20:0: R0903: Too few public methods (1/2) (too-few-public-methods)
> timeless/restaurants/models.py:33:0: R0903: Too few public methods (1/2) (too-few-public-methods)
> timeless/restaurants/models.py:48:0: R0903: Too few public methods (1/2) (too-few-public-methods)
> timeless/restaurants/models.py:76:0: R0903: Too few public methods (0/2) (too-few-public-methods)
> timeless/restaurants/models.py:88:0: R0903: Too few public methods (1/2) (too-few-public-methods)
> timeless/restaurants/models.py:118:0: R0903: Too few public methods (1/2) (too-few-public-methods)
> ************* Module timeless.restaurants.tables.views
> timeless/restaurants/tables/views.py:23:0: C0111: Missing function docstring (missing-docstring)
> timeless/restaurants/tables/views.py:59:0: C0111: Missing function docstring (missing-docstring)
> timeless/restaurants/tables/views.py:77:0: C0111: Missing function docstring (missing-docstring)
> timeless/restaurants/tables/views.py:77:9: W0613: Unused argument 'id' (unused-argument)
> timeless/restaurants/tables/views.py:95:0: C0111: Missing function docstring (missing-docstring)
> ************* Module timeless.restaurants.floors.views
> timeless/restaurants/floors/views.py:43:9: W0613: Unused argument 'id' (unused-argument)
> ************* Module timeless.restaurants.table_shapes.forms
> timeless/restaurants/table_shapes/forms.py:6:0: W0223: Method 'update' is abstract in class 'ModelForm' but is not overridden (abstract-method)
> timeless/restaurants/table_shapes/forms.py:9:4: C0111: Missing class docstring (missing-docstring)
> timeless/restaurants/table_shapes/forms.py:9:4: R0903: Too few public methods (0/2) (too-few-public-methods)
> ************* Module timeless.restaurants.table_shapes.views
> timeless/restaurants/table_shapes/views.py:39:-1: W0105: String statement has no effect (pointless-string-statement)
> timeless/restaurants/table_shapes/views.py:47:9: W0613: Unused argument 'id' (unused-argument)
> timeless/restaurants/table_shapes/views.py:65:4: E1101: Instance of 'scoped_session' has no 'delete' member (no-member)
> ************* Module timeless.restaurants.table_shapes.__init__
> timeless/restaurants/table_shapes/__init__.py:1:0: R0801: Similar lines in 2 files
> ==timeless.restaurants.tables.views:72
> ==timeless.roles.views:43
>     )
> 
> 
> @bp.route("/edit/<int:id>", methods=("GET", "POST"))
> def edit(id):
>     if request.method == "POST":
>         flash("Edit not yet implemented") (duplicate-code)
> 
> -----------------------------------
> Your code has been rated at 7.98/10
> 
> container 801c354a71927eafe41aed203c5dda2a9638c16f252834db9d77aacd65cc5caa is dead
> Sun Feb 10 21:23:55 CET 2019
> ```

@emilianodellacasa 
I guess this is not a problem with my PR right?

**Comment by emilianodellacasa** [comment]
@ammaratef45 maybe not, but we cannot merge. A simple fix to a pylint warning should be enough

**Comment by ammaratef45** [comment]
What should I do?

Sent from Yahoo Mail on Android 
 
  On Sun, Feb 10, 2019 at 10:35 PM, Emiliano Della Casa<notifications@github.com> wrote:   
@ammaratef45 maybe not, but we cannot merge. A simple fix to a pylint warning should be enough

**Comment by vladarefiev** [comment]
@amihaiemil First of all fix puzzle message, length should be at least 20 words otherwise it will fail on the next step
Secondly for some reason pylint has rated code as 7.98, bit 8.0 is min. Fix any of warnings and it will be passed

**Comment by ammaratef45** [comment]
@vladarefiev you meant to mention me not @amihaiemil right?

**Comment by vladarefiev** [comment]
@ammaratef45 My bad :)
I want to mention you

**Comment by ammaratef45** [comment]
@vladarefiev Most of the lint errors are not mine, I reported #195 before for this.
Now we can:
   -Edit some in a new commit (not relevant to the issue) to make the merge happens
   -Wait till #195 be solved
   -Tell linter to ignore some of them for now.

**Comment by emilianodellacasa** [comment]
@ammaratef45 A couple of them are actually yours:

```
timeless/__init__.py:19:0: C0301: Line too long (105/100) (line-too-long)
timeless/__init__.py:12:0: C0411: third party import "from flask_caching import Cache" should be placed before "from timeless.db import DB" (wrong-import-order)
```

**Comment by ammaratef45** [comment]
The score is now 8.05, the todo is longer.
5afc2eb is my last commit.
The jam of lint warnings makes it hard to catch what is mine.
I hope in the next PR's we can improve it by fixing lint warnings that are in the same files we edit. :+1:

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@ammaratef45 A small error in your puzzle

**Comment by ammaratef45** [comment]
> @ammaratef45 A small error in your puzzle

I guess it is not my puzzle
pdd found a wrong todo in the redis source code downloaded in the installation process by rultor.
That won't happen if I installed redis in the parent directory of the project? I will try

**Comment by emilianodellacasa** [comment]
@ammaratef45 I know, trying again after excluding that dir

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by ammaratef45** [comment]
Oh, it's not even a todo
this character ¥ is in the line PDD mentioned
should be a pug with PDD

**Comment by ammaratef45** [comment]
@emilianodellacasa looks like you that dir isn't excluded.
should I try changing the Redis installation script?

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#164](https://github.com/timelesslounge/timelessis/issues/164) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@vladarefiev ccording to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only two issues were found during code review. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by ypshenychka** [comment]
@ammaratef45 According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>Messages in a job [always](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html#4-avoid-noiseaddress-your-comments) start with a name of a user they are addressed to.

Many of your messages are not addressed. Please go through your messages and correct them by indicating an addressee in the beginning.

**Comment by vladarefiev** [comment]
@ypshenychka correct

**Comment by ypshenychka** [comment]
@vladarefiev thanks

**Comment by ammaratef45** [comment]
@ypshenychka Thanks
I edited my messages, you can take a look.

**Comment by ypshenychka** [comment]
@ammaratef45 thank you

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @vladarefiev[/z](https://www.0crat.com/u/vladarefiev)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**